api/oidc.go

```diff
@@ -0,0 +1,32 @@
+package api
+
+import (
+	"fmt"
+)
+
+type OidcToken struct {
+	Token string `json:"token"`
+}
+
+type OidcTokenRequest struct {
+	JobId    string
+	Audience string
+}
+
+func (c *Client) OidcToken(methodReq *OidcTokenRequest) (*OidcToken, *Response, error) {
+	m := &struct {
+		Audience string `json:"audience,omitempty"`
+	}{
+		Audience: methodReq.Audience,
+	}
+
+	u := fmt.Sprintf("jobs/%s/oidc/tokens", methodReq.JobId)
+	httpReq, err := c.newRequest("POST", u, m)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	t := &OidcToken{}
+	resp, err := c.doRequest(httpReq, t)
```

Comment:
Are the token and response ever useful if err != nil? If so that would make using this method more complicated.
Otherwise to avoid potential confusion ("I called OIDCToken and got both an error and some other stuff?") I would insert an `if err != nil { return nil, nil, err }` in here.

Reply:
`doRequest` returns the response in cases where `err != nil`. The rationale seems to be that it allows inspection of the status code and retrying if necessary. See https://github.com/buildkite/agent/blob/1a9d87e123f3b2c07c86d763a05e870a6395da0c/api/client.go#L260-L261

The practice in this codebase seems to be to do the retrying in the `clicommand` method and not the `api` one, and inspecting the status code from the bubbled up response. See https://github.com/buildkite/agent/blob/1a9d87e123f3b2c07c86d763a05e870a6395da0c/api/meta_data.go#L60-L65 and https://github.com/buildkite/agent/blob/1a9d87e123f3b2c07c86d763a05e870a6395da0c/clicommand/meta_data_get.go#L107-L119

So the respone SHOULD be passed up or the retry won't work as expected. The token is not needed not however, so I have changed it to pass a nil pointer on error now.

I don't really think this is a good practice though (see below). However, I do think it's better to consistently follow a bad practice than to mix different practices. So I think we should follow this for now, and have a piece of work to unravel `http.doRequest`. I haven't thought too deeply about what that should be, but here are some initial thoughts:

I think there is no good reason for `api.doRequest` to return an `http.Response`. Instead, I think the relavant parts of the response that are needed for retry handling or otherwise should be returned as part of a structured error. Indeed, there already is a error type that purports to do this: `api.ErrorReponse`, although it contains a pointer to `http.Response` 🤷‍♂️. Either way returning a pointer to `http.Response` is potentially misleading as the body of the `http.Response` is closed in `api.doRequest`, so it will look like there is no body if we try to use it outside the `api.doRequest` method.

Reply:
That's fair. `net/http` itself does this sort of thing.